x-pack/plugins/ml/public/url_generator.ts

```diff
@@ -64,39 +208,265 @@ export class MlUrlGenerator implements UrlGeneratorsDefinition<typeof ML_APP_URL
 
   public readonly id = ML_APP_URL_GENERATOR;
 
-  public readonly createUrl = async ({ page, ...params }: MlUrlGeneratorState): Promise<string> => {
-    if (page === 'explorer') {
-      return this.createExplorerUrl(params);
+  public readonly createUrl = async (mlUrlGeneratorState: MlUrlGeneratorState): Promise<string> => {
+    switch (mlUrlGeneratorState.page) {
+      case ML_PAGES.ANOMALY_DETECTION:
+        return this.createAnomalyDetectionJobManagementUrl(
+          mlUrlGeneratorState as AnomalyDetectionUrlState
+        );
+      case ML_PAGES.ANOMALY_EXPLORER:
+        return this.createExplorerUrl(mlUrlGeneratorState as ExplorerUrlState);
+      case ML_PAGES.TIME_SERIES_EXPLORER:
+        return this.createSingleMetricViewerUrl(mlUrlGeneratorState as TimeSeriesExplorerUrlState);
+      case ML_PAGES.DATA_FRAME_ANALYTICS:
+        return this.createDataframeAnalyticsUrl(mlUrlGeneratorState as DataFrameAnalyticsUrlState);
+      case ML_PAGES.DATA_FRAME_ANALYTICS_EXPLORATION:
+        return this.createDataframeAnalyticsExplorationUrl(
+          mlUrlGeneratorState as DataFrameAnalyticsExplorationUrlState
+        );
+      case ML_PAGES.DATA_VISUALIZER:
+        return this.createDataVisualizerUrl(mlUrlGeneratorState as DataVisualizerUrlState);
+      case ML_PAGES.DATA_VISUALIZER_FILE:
+        return this.createDataVisualizerUrl(mlUrlGeneratorState as DataVisualizerUrlState);
+      case ML_PAGES.DATA_VISUALIZER_INDEX_SELECT:
+        return this.createDataVisualizerUrl(mlUrlGeneratorState as DataVisualizerUrlState);
+      case ML_PAGES.DATA_VISUALIZER_NEW_JOB:
+        return this.createIndexBasedMlUrl(mlUrlGeneratorState as MlGenericUrlState);
+      case ML_PAGES.DATA_VISUALIZER_VIEWER:
+        return this.createIndexBasedMlUrl(mlUrlGeneratorState as MlGenericUrlState);
```

Comment:
Is this correct? It calls the same `createIndexBasedMlUrl` as the `DATA_VISUALIZER_NEW_JOB` case above?

Reply:
This is correct, `createIndexBasedMlUrl` was intended to be generic here so it can work with both `DATA_VISUALIZER_INDEX_VIEWER` and `ANOMALY_DETECTION_CREATE_JOB_SELECT_TYPE`. Since they are fairly similar (both have either `index` or `savedSearchId`, `_g`, and `_a`), I thought using this.createIndexBasedMlUrl should be fine since the functions are private anyway.

However, I can see why it's confusing. I'll create new functions that wrap around `this.createIndexBasedMlUrl`.

Reply:
Updated here https://github.com/elastic/kibana/pull/75696/commits/49e56edc23a2a9ddc12e9145add11bba8f2b2e7f

Although I'm also happy to have the new functions do their own dedicated thing without calling `createIndexBasedMlUrl`